From 994de10daafca166b2a62dcb86eae99a1fcdb68a Mon Sep 17 00:00:00 2001
From: Jeremy Huddleston Sequoia <jeremyhu@apple.com>
Date: Fri, 27 Jan 2023 16:42:29 -0800
Subject: [PATCH 01/18] os: Add an implementation of wl_os_socket_peercred()
 for darwin

This adds an alternative implementation of wl_os_socket_peercred() using
getpeereid(3) and LOCAL_PEERPID, similar to changes I recently made to
xorg-server.

Signed-off-by: Jeremy Huddleston Sequoia <jeremyhu@apple.com>
---
 meson.build      |  1 +
 src/wayland-os.c | 16 ++++++++++++++++
 2 files changed, 17 insertions(+)

diff --git a/meson.build b/meson.build
index ef34ba4..9fd5416 100644
--- a/meson.build
+++ b/meson.build
@@ -40,6 +40,7 @@ endforeach
 
 have_funcs = [
 	'accept4',
+	'getpeereid',
 	'mkostemp',
 	'posix_fallocate',
 	'prctl',
diff --git a/src/wayland-os.c b/src/wayland-os.c
index f00ead4..e134c3a 100644
--- a/src/wayland-os.c
+++ b/src/wayland-os.c
@@ -125,6 +125,22 @@ wl_os_socket_peercred(int sockfd, uid_t *uid, gid_t *gid, pid_t *pid)
 	*pid = ucred.pid;
 	return 0;
 }
+#elif defined(HAVE_GETPEEREID) && defined(LOCAL_PEERPID)
+int
+wl_os_socket_peercred(int sockfd, uid_t *uid, gid_t *gid, pid_t *pid)
+{
+	socklen_t len;
+
+	if (getpeereid(sockfd, uid, gid) != 0) {
+		return -1;
+	}
+
+	len = sizeof(pid_t);
+	if (getsockopt(sockfd, SOL_LOCAL, LOCAL_PEERPID, pid, &len) != 0) {
+		return -1;
+	}
+	return 0;
+}
 #else
 #error "Don't know how to read ucred on this platform"
 #endif
-- 
2.24.3 (Apple Git-128)

